### ISSUE TITLE
⚡ Report bugs and suggest features for v2 here

### ISSUE THREAD

**Issue by balkarjun** [jarbun/downline#15]
🎉 After 3 months of development, the first alpha release of v2.0 of downline is here (for windows)
 
![screenshot](https://user-images.githubusercontent.com/23068820/67160435-40c33d80-f36e-11e9-8e44-296cbb20dc69.png)

[Download Latest](https://github.com/jarbun/downline/releases)

Incomplete and probably hella buggy, but still a significant improvement over the previous version. A detailed changelog and list of planned features will be added later.

The development stages will be as follows: 
- **Alpha** - With a focus on adding features and fixing bugs
- **Beta** - UI and bug fixes
- **Stable Release**

## This thread is for reporting bugs, suggesting features and asking questions related to the alpha release

**Comment by stefnotch** [comment]
Great work! 😄 I just tried it out and here is my first feedback.
- [x]  I'm a big fan of the new filename settings
- [ ] When you input a link to a private playlist or a nonexistent video, the downloader does nothing. It'd be preferable to have a little message about which links failed.
- [ ] When clicking on clear all, the dialog that pops up seems to have a dark theme, which is odd.
- [ ] I'm a fan of the flat UI, but none of the buttons change color or have any indicator that you can press them, other than the cursor changing. Although, I suppose such CSS things will be addressed in future updates.
- [ ] Downline understandably doesn't handle playlist reordering perfectly. 
![image](https://user-images.githubusercontent.com/10220080/67623651-1c60d880-f828-11e9-86ef-2e7226011716.png)
Steps to reproduce:
1. Change the settings to autonumber the videos
2. Download a playlist with a few videos
3. Remove a few videos from the list
4. Remove them from the list
5. Change their playlist index on YouTube
6. Download them again. Now they have a new index.
In my opinion, it would be preferable if Downline simply renamed the files if it already downloaded that video. (And `Download All` should do that with all files)

**Comment by balkarjun** [comment]
> When you input a link to a private playlist or a nonexistent video, the downloader does nothing. It'd be preferable to have a little message about which links failed.

Better error handling will be added soon. I'll make sure to address these issues.

>  the dialog that pops up seems to have a dark theme

This was intentional, and meant to test whether a light theme (such as the dropdown when selecting quality) or dark theme was better. 
It will be made consistent later, so I'd like to know what you think. Should I change all dialogs and dropdowns to a dark theme (for improved contrast) or change them all to the light theme (for better consistency with the background)?

> none of the buttons change color or have any indicator that you can press them

Most of the UI work will be done in the beta releases. The focus right now is on bug fixes and feature additions, so the UI is nowhere near completion. Improved button states, among other UI work, will be added later

>  it would be preferable if Downline simply renamed the files if it already downloaded that video

This would fix the issue, but I'm not sure this is a good solution (If i understood correctly. Please correct me if I made the wrong assumptions). Consider the following cases:
1. Playlist with videos that have same names. If video X and video Y have the same name,  and X is downloaded, then this implementation would cause Y to not be downloaded and instead rename X (unless id or title-id is chosen in filename options). This could be fixed by keeping track of downloaded files internally somewhow, but I'm not sure if that would be a good solution.
2. User wants to re-download a playlist item with different settings (quality, audio, subtitle etc).

**Comment by stefnotch** [comment]
> This was intentional, and meant to test whether a light theme (such as the dropdown when selecting quality) or dark theme was better.
It will be made consistent later, so I'd like to know what you think. Should I change all dialogs and dropdowns to a dark theme (for improved contrast) or change them all to the light theme (for better consistency with the background)?

I prefer it if everything has a consistent theme. In the case of a dialog, the background could be slightly darkened.

> This would fix the issue, but I'm not sure this is a good solution (If i understood correctly. Please correct me if I made the wrong assumptions). Consider the following cases:
> 1. Playlist with videos that have same names ...
> 2. User wants to re-download a playlist item ...

Oh, good points. I didn't consider them, because I wrongly assumed that Downline internally keeps track of the video IDs. I'm not sure what the best solution would be.

**Comment by stefnotch** [comment]
I have a question regarding playlist downloading: Why does it take so long to enumerate the entire playlist? (e.g. I have a playlist with ~100 videos and sometimes add a new video to it) Is that a youtube-dl limitation?

**Comment by balkarjun** [comment]
@stefnotch Yes, it is a limitation of youtube-dl. You can verify this by executing the following command, which is what Downline uses to fetch video info:

```youtube-dl.exe --all-subs --dump-json --no-playlist --ignore-errors <playlist-url>```

**Comment by BuyMyMojo** [comment]
Feature request:
option to use hardware decode+encode for converting to other file formats for nvidia, amd and intel enabled devices?

**Comment by stefnotch** [comment]
Feature request:
If a user has `ffmpeg` and/or `youtube-dl` installed, it'd be neat if Downline could use those instead of manually having to download them.

**Comment by stefnotch** [comment]
Another feature I'd love to see would be better handling of huge playlists. For example, by getting a large playlist in chunks of 50 videos. I have a playlist with ~100 videos and youtube-dl hangs at 91.
![image](https://user-images.githubusercontent.com/10220080/85957078-8e2a8c80-b98a-11ea-8822-b3de615eff39.png)

**Update:** It actually managed to finish, however, one video is missing.

**Comment by stefnotch** [comment]
@jarbun I might have some free time on my hands and would love to contribute to this project. What would be the best way to get started and is there anything I really should take a look at? One thing I want to do is improving the handling of large playlists.

**Comment by stefnotch** [comment]
Feature suggestion: Support `--embed-thumbnail`
